client/test_commands/test_bw6_761_groth16_contract.py

```diff
@@ -0,0 +1,81 @@
+# Copyright (c) 2015-2020 Clearmatics Technologies Ltd
+#
+# SPDX-License-Identifier: LGPL-3.0+
+
+from zecale.core.utils import get_zecale_dir
+from zecale.cli.utils import load_verification_key, load_transaction
+from zeth.core.zksnark import Groth16SnarkProvider, GenericVerificationKey
+from zeth.core.utils import hex_list_to_uint256_list
+from zeth.core.contracts import InstanceDescription
+from zeth.cli.utils import get_eth_network, open_web3_from_network
+from os.path import join
+import sys
+from typing import Any
+
+
+ZECALE_DIR = get_zecale_dir()
+CONTRACTS_DIR = join(ZECALE_DIR, "contracts")
+DUMMY_APP_DIR = join(ZECALE_DIR, "testdata", "dummy_app")
+
+
+def _test_bw6_761_groth16_contract_with_proof(
+        instance: Any,
+        vk: GenericVerificationKey,
+        proof_filename: str) -> bool:
+    # Load proof and extract inputs
+    extproof = load_transaction(join(DUMMY_APP_DIR, proof_filename))
+    inputs = extproof["inputs"]
+
+    # Encode the vk, proof and inputs into evm words
+    zksnark = Groth16SnarkProvider()
+    vk_evm_parameters = zksnark.verification_key_to_contract_parameters(vk)
+    proof_evm_parameters = zksnark.proof_to_contract_parameters(extproof)
+    inputs_evm_parameters = hex_list_to_uint256_list(inputs)
+
+    # Execute the test contract and return the result
+    evm_parameters = [
+        vk_evm_parameters,
+        proof_evm_parameters,
+        inputs_evm_parameters
+    ]
+    return instance.functions.test_verify(*evm_parameters).call()
+
+
+def test_bw6_761_groth16_valid(
+        instance: Any, aggregator_vk: GenericVerificationKey) -> None:
+    assert _test_bw6_761_groth16_contract_with_proof(
+        instance, aggregator_vk, "batch1.json")
+
+
+def test_bw6_761_groth16_invalid(
+        instance: Any, aggregator_vk: GenericVerificationKey) -> None:
+    assert not _test_bw6_761_groth16_contract_with_proof(
+        instance, aggregator_vk, "batch1-invalid.json")
+
+
+def main() -> int:
+    web3: Any = open_web3_from_network(get_eth_network(None))
+    bw6_761_groth16_instance_desc = InstanceDescription.deploy(
+        web3,
+        join(CONTRACTS_DIR, "bw6_761_groth16_test.sol"),
+        "bw6_761_groth16_test",
+        web3.eth.accounts[0],  # pylint: disable=no-member
+        None,
+        500000,
```

Comment:
Can't we rely on `estimateGas` here?

Reply:
Related to https://github.com/clearmatics/zeth/issues/295

Reply:
For now, this would require a change to the contract infrastructure in Zeth.  Not complex, but probably easier to handle separately.

Reply:
Ah right, this is the deployment, not a call (gas isn't optional for the deployment https://github.com/clearmatics/zeth/blob/develop/client/zeth/core/contracts.py#L135 but is for calls e.g. https://github.com/clearmatics/zeth/blob/develop/client/zeth/core/contracts.py#L345)
Let's keep it that way then, but that would be nice to make sure that gas is optional (to use `estimateGas`) for the Zecale contract calls going forward :)